File: shared/python/engine/math_ops.py
"""
Logic Engine - Math Operations

Pure functions for math operations.
Uses integer arithmetic to match C implementation.
"""

from enum import IntEnum
from typing import Sequence


class MathOp(IntEnum):
    """Math operation types"""
    ADD = 0
    SUB = 1
    MUL = 2
    DIV = 3
    MOD = 4
    ABS = 5
    NEG = 6
    MIN = 16
    MAX = 17
    AVG = 18
    CLAMP = 32
    MAP = 33
    SCALE = 34
    LERP = 35


# Constants matching C implementation
INT32_MAX = 2147483647
INT32_MIN = -2147483648


def _clamp_i32(value: int) -> int:
    """Clamp to int32 range"""
    if value > INT32_MAX:
        return INT32_MAX
    if value < INT32_MIN:
        return INT32_MIN
    return value


def math_add(a: int, b: int) -> int:
    """Add two values with overflow protection"""
    return _clamp_i32(a + b)


def math_sub(a: int, b: int) -> int:
    """Subtract two values with overflow protection"""
    return _clamp_i32(a - b)


def math_mul(a: int, b: int) -> int:
    """Multiply two values with overflow protection"""
    return _clamp_i32(a * b)


def math_div(a: int, b: int) -> int:
    """Divide two values (returns 0 on division by zero)"""
    if b == 0:
        return INT32_MAX if a > 0 else (INT32_MIN if a < 0 else 0)
    return a // b


def math_mod(a: int, b: int) -> int:
    """Modulo operation (returns 0 on division by zero)"""
    if b == 0:
        return 0
    return a % b


def math_abs(value: int) -> int:
    """Absolute value"""
    if value == INT32_MIN:
        return INT32_MAX  # Prevent overflow
    return abs(value)


def math_neg(value: int) -> int:
    """Negate value"""
    if value == INT32_MIN:
        return INT32_MAX
    return -value


def math_min(inputs: Sequence[int]) -> int:
    """Find minimum of all inputs"""
    if not inputs:
        return 0
    return min(inputs)


def math_max(inputs: Sequence[int]) -> int:
    """Find maximum of all inputs"""
    if not inputs:
        return 0
    return max(inputs)


def math_avg(inputs: Sequence[int]) -> int:
    """Calculate average of all inputs (integer division)"""
    if not inputs:
        return 0
    return sum(inputs) // len(inputs)


def math_clamp(value: int, min_val: int, max_val: int) -> int:
    """Clamp value to range [min_val, max_val]"""
    if value < min_val:
        return min_val
    if value > max_val:
        return max_val
    return value


def math_map(value: int, in_min: int, in_max: int, out_min: int, out_max: int) -> int:
    """
    Map value from input range to output range.

    Equivalent to Arduino map() function.
    Uses 64-bit intermediate to prevent overflow.
    """
    in_range = in_max - in_min
    if in_range == 0:
        return out_min

    out_range = out_max - out_min
    result = (value - in_min) * out_range // in_range + out_min

    return _clamp_i32(result)


def math_scale(value: int, multiplier: int, divisor: int) -> int:
    """
    Scale value by multiplier/divisor.
    Uses 64-bit intermediate to prevent overflow.
    """
    if divisor == 0:
        return INT32_MAX if value * multiplier > 0 else INT32_MIN
    return _clamp_i32(value * multiplier // divisor)


def math_lerp(a: int, b: int, t: int, t_max: int = 1000) -> int:
    """
    Linear interpolation between a and b.
    t=0 returns a, t=t_max returns b.
    Default t_max=1000 for 0.1% resolution.
    """
    if t_max == 0:
        return a
    if t <= 0:
        return a
    if t >= t_max:
        return b

    return _clamp_i32(a + (b - a) * t // t_max)


def math_evaluate(op: MathOp, inputs: Sequence[int]) -> int:
    """
    Evaluate math operation.

    Args:
        op: Math operation to perform
        inputs: Input values (meaning depends on operation)

    Returns:
        Operation result
    """
    if not inputs:
        return 0

    if op == MathOp.ADD:
        return math_add(inputs[0], inputs[1]) if len(inputs) >= 2 else inputs[0]
    elif op == MathOp.SUB:
        return math_sub(inputs[0], inputs[1]) if len(inputs) >= 2 else inputs[0]
    elif op == MathOp.MUL:
        return math_mul(inputs[0], inputs[1]) if len(inputs) >= 2 else inputs[0]
    elif op == MathOp.DIV:
        return math_div(inputs[0], inputs[1]) if len(inputs) >= 2 else inputs[0]
    elif op == MathOp.MOD:
        return math_mod(inputs[0], inputs[1]) if len(inputs) >= 2 else inputs[0]
    elif op == MathOp.ABS:
        return math_abs(inputs[0])
    elif op == MathOp.NEG:
        return math_neg(inputs[0])
    elif op == MathOp.MIN:
        return math_min(inputs)
    elif op == MathOp.MAX:
        return math_max(inputs)
    elif op == MathOp.AVG:
        return math_avg(inputs)
    elif op == MathOp.CLAMP:
        return math_clamp(inputs[0], inputs[1], inputs[2]) if len(inputs) >= 3 else inputs[0]
    elif op == MathOp.MAP:
        return math_map(inputs[0], inputs[1], inputs[2], inputs[3], inputs[4]) if len(inputs) >= 5 else inputs[0]
    elif op == MathOp.SCALE:
        return math_scale(inputs[0], inputs[1], inputs[2]) if len(inputs) >= 3 else inputs[0]
    elif op == MathOp.LERP:
        if len(inputs) >= 4:
            return math_lerp(inputs[0], inputs[1], inputs[2], inputs[3])
        elif len(inputs) >= 3:
            return math_lerp(inputs[0], inputs[1], inputs[2])
        return inputs[0]

    return 0

File: shared/engine/math_ops.c
/**
 * @file math_ops.c
 * @brief Logic Engine - Math Operations Implementation
 */

#include "math_ops.h"

/*============================================================================
 * Basic Arithmetic
 *============================================================================*/

int32_t Math_Add(const int32_t* inputs, uint8_t count)
{
    if (!inputs || count == 0) return 0;

    int32_t sum = 0;
    for (uint8_t i = 0; i < count; i++) {
        sum += inputs[i];
    }
    return sum;
}

int32_t Math_Sub(int32_t a, int32_t b)
{
    return a - b;
}

int32_t Math_Mul(int32_t a, int32_t b)
{
    return a * b;
}

int32_t Math_Div(int32_t a, int32_t b)
{
    if (b == 0) return 0;  /* Safe division */
    return a / b;
}

int32_t Math_Mod(int32_t a, int32_t b)
{
    if (b == 0) return 0;  /* Safe modulo */
    return a % b;
}

int32_t Math_Abs(int32_t value)
{
    return (value < 0) ? -value : value;
}

int32_t Math_Neg(int32_t value)
{
    return -value;
}

/*============================================================================
 * Min/Max/Average
 *============================================================================*/

int32_t Math_Min(const int32_t* inputs, uint8_t count)
{
    if (!inputs || count == 0) return 0;

    int32_t min = inputs[0];
    for (uint8_t i = 1; i < count; i++) {
        if (inputs[i] < min) min = inputs[i];
    }
    return min;
}

int32_t Math_Max(const int32_t* inputs, uint8_t count)
{
    if (!inputs || count == 0) return 0;

    int32_t max = inputs[0];
    for (uint8_t i = 1; i < count; i++) {
        if (inputs[i] > max) max = inputs[i];
    }
    return max;
}

int32_t Math_Avg(const int32_t* inputs, uint8_t count)
{
    if (!inputs || count == 0) return 0;

    int32_t sum = Math_Add(inputs, count);
    return sum / (int32_t)count;
}

/*============================================================================
 * Range Operations
 *============================================================================*/

int32_t Math_Clamp(int32_t value, int32_t min, int32_t max)
{
    if (value < min) return min;
    if (value > max) return max;
    return value;
}

int32_t Math_Map(int32_t value, int32_t in_min, int32_t in_max,
                 int32_t out_min, int32_t out_max)
{
    /* Avoid division by zero */
    int32_t in_range = in_max - in_min;
    if (in_range == 0) return out_min;

    /* Calculate mapped value using 64-bit to avoid overflow */
    int64_t out_range = (int64_t)(out_max - out_min);
    int64_t value_offset = (int64_t)(value - in_min);

    return out_min + (int32_t)((value_offset * out_range) / in_range);
}

int32_t Math_Scale(int32_t value, int32_t multiplier, int32_t scale_factor)
{
    if (scale_factor == 0) return 0;

    /* Use 64-bit to avoid overflow */
    int64_t result = ((int64_t)value * multiplier) / scale_factor;

    /* Clamp to int32 range */
    if (result > INT32_MAX) return INT32_MAX;
    if (result < INT32_MIN) return INT32_MIN;

    return (int32_t)result;
}

int32_t Math_Lerp(int32_t a, int32_t b, int32_t t)
{
    /* t is 0-1000 (representing 0.0 to 1.0) */
    if (t <= 0) return a;
    if (t >= 1000) return b;

    /* a + (b - a) * t / 1000 */
    int64_t diff = (int64_t)(b - a);
    return a + (int32_t)((diff * t) / 1000);
}

/*============================================================================
 * Generic Evaluate Function
 *============================================================================*/

int32_t Math_Evaluate(MathOp_t op, const int32_t* inputs, uint8_t count)
{
    if (!inputs || count == 0) return 0;

    switch (op) {
        case MATH_OP_ADD:
            return Math_Add(inputs, count);

        case MATH_OP_SUB:
            return (count >= 2) ? Math_Sub(inputs[0], inputs[1]) : inputs[0];

        case MATH_OP_MUL:
            return (count >= 2) ? Math_Mul(inputs[0], inputs[1]) : inputs[0];

        case MATH_OP_DIV:
            return (count >= 2) ? Math_Div(inputs[0], inputs[1]) : inputs[0];

        case MATH_OP_MOD:
            return (count >= 2) ? Math_Mod(inputs[0], inputs[1]) : inputs[0];

        case MATH_OP_ABS:
            return Math_Abs(inputs[0]);

        case MATH_OP_NEG:
            return Math_Neg(inputs[0]);

        case MATH_OP_MIN:
            return Math_Min(inputs, count);

        case MATH_OP_MAX:
            return Math_Max(inputs, count);

        case MATH_OP_AVG:
            return Math_Avg(inputs, count);

        case MATH_OP_CLAMP:
            /* inputs[0]=value, inputs[1]=min, inputs[2]=max */
            return (count >= 3) ? Math_Clamp(inputs[0], inputs[1], inputs[2]) : inputs[0];

        case MATH_OP_MAP:
            /* inputs[0]=value, [1]=in_min, [2]=in_max, [3]=out_min, [4]=out_max */
            return (count >= 5) ?
                Math_Map(inputs[0], inputs[1], inputs[2], inputs[3], inputs[4]) : inputs[0];

        case MATH_OP_SCALE:
            /* inputs[0]=value, inputs[1]=multiplier, inputs[2]=scale_factor */
            return (count >= 3) ?
                Math_Scale(inputs[0], inputs[1], inputs[2]) : inputs[0];

        default:
            return 0;
    }
}
